source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,927 @@
+              // is only one value to be copied, but for non-primitive
+              // elements, there might be more values to be copied
+              //
+              // so, find out from which index to take this one value, and
+              // to which index to put
+              unsigned int out_index = 0;
+              for (unsigned int i=0; i<system_index; ++i)
+                out_index += this->n_nonzero_components(i);
+              unsigned int in_index = 0;
+              for (unsigned int i=0; i<base_index; ++i)
+                in_index += base_fe.n_nonzero_components(i);
+
+              // then loop over the number of components to be copied
+              Assert (this->n_nonzero_components(system_index) ==
+                      base_fe.n_nonzero_components(base_index),
+                      ExcInternalError());
+              for (unsigned int s=0; s<this->n_nonzero_components(system_index); ++s)
+                {
+                  if (base_flags & update_values)
+                    for (unsigned int q=0; q<n_q_points; ++q)
+                      output_data.shape_values[out_index+s][q] =
+                        base_data.shape_values(in_index+s,q);
+
+                  if (base_flags & update_gradients)
+                    for (unsigned int q=0; q<n_q_points; ++q)
+                      output_data.shape_gradients[out_index+s][q] =
+                        base_data.shape_gradients[in_index+s][q];
+
+                  if (base_flags & update_hessians)
+                    for (unsigned int q=0; q<n_q_points; ++q)
+                      output_data.shape_hessians[out_index+s][q] =
+                        base_data.shape_hessians[in_index+s][q];
+
+                }
+            }
+      }
+
+  Assert (base_no_mult_local_enriched_dofs.size() == fe_data.enrichment.size(),
+          ExcDimensionMismatch(base_no_mult_local_enriched_dofs.size(),
+                               fe_data.enrichment.size()));
+  // calculate hessians, gradients and values for each function
+  for (unsigned int base_no = 1; base_no < this->n_base_elements(); base_no++)
+    {
+      Assert (base_no_mult_local_enriched_dofs[base_no].size() == fe_data.enrichment[base_no].size(),
+              ExcDimensionMismatch(base_no_mult_local_enriched_dofs[base_no].size(),
+                                   fe_data.enrichment[base_no].size()));
+      for (unsigned int m=0; m < base_no_mult_local_enriched_dofs[base_no].size(); m++)
+        {
+          if (flags & update_hessians)
+            {
+              Assert (fe_data.enrichment[base_no][m].hessians.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].hessians.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].hessians[q]  = enrichments[base_no-1][m](cell)->hessian (mapping_data.quadrature_points[q]);
```

Comment:
here and in the same lines in the following two blocks, you call `enrichments[...][m](cell)` which gives you a `Function<dim>*`. I would like it if there was a check somewhere that this is really a scalar function. E.g., at the top of the loop over `m`, you could have
```
  Assert (enrichments[base_no-1][m](cell)->n_components() == 1, ...);
```